aiida/manage/configuration/migrations/migrations.py

```diff
@@ -8,136 +8,282 @@
 # For further information please visit http://www.aiida.net               #
 ###########################################################################
 """Define the current configuration version and migrations."""
+from typing import Any, Dict, Protocol
 
-__all__ = ('CURRENT_CONFIG_VERSION', 'OLDEST_COMPATIBLE_CONFIG_VERSION')
+from aiida.common import exceptions
+
+__all__ = (
+    'CURRENT_CONFIG_VERSION', 'OLDEST_COMPATIBLE_CONFIG_VERSION', 'get_current_version', 'check_and_migrate_config',
+    'config_needs_migrating', 'upgrade_config', 'downgrade_config'
+)
+
+ConfigType = Dict[str, Any]
 
 # The expected version of the configuration file and the oldest backwards compatible configuration version.
 # If the configuration file format is changed, the current version number should be upped and a migration added.
 # When the configuration file format is changed in a backwards-incompatible way, the oldest compatible version should
 # be set to the new current version.
+
 CURRENT_CONFIG_VERSION = 5
 OLDEST_COMPATIBLE_CONFIG_VERSION = 5
 
 
-class ConfigMigration:
-    """Defines a config migration."""
+class SingleMigration(Protocol):
+    """A single migration of the configuration."""
+    initial: int
+    """The initial configuration version."""
```

Comment:
```suggestion
    initial: int  #The initial configuration version.
```

Reply:
No, you shouldn't do it like this; triple-back tick comments, under a variable, are recognised correctly as its description, and shown for example in sphinx documentation and LSPs:

![image](https://user-images.githubusercontent.com/2997570/150364507-7a5e6af8-e898-48b2-b17d-5c324811eb8a.png)

Reply:
I see your PEP8 and raise you https://www.python.org/dev/peps/pep-0224/

Reply:
> "String literals occurring immediately after a simple assignment at the top level of a module, class, or `__init__`` method are called "attribute docstrings"."

Reply:
so yeh, we should be using it

Reply:
That phrase is the only reference to it and they don't even give an example. There was also a good reason to reject the concept, because it is just not clear to which attribute the docstring belongs to. This is also the reason I stumbled over it:
```python
initial: int
"""Some docstring"""
final: int
```
It is just not logical to think that the docstring belongs to `initial` and not `final`.

Guess in the end it is just personal preference and here I would again consider readability. The fact that certain software tools have added support for it, should not always be sufficient reason to adopt it if the majority of times that the code is read, is without interpretation. But I will let others give their opinion and not fight this if others think it is fine.

Reply:
Its implemented in sphinx and most every Python LSP, it's not just personnel preference. 

> If the majority of times that the code is read, is without interpretation.

I disagree this is the case, and I would certainly recommend you to use  a modern LSP in your development

Reply:
I agree, but in those cases there is no room for confusion since there are never two consecutive declarations without any whitespace. I think it is that then that distinguishes the cases really, at least put a line of whitespace between the docstring and the next attribute declaration. This should make it easier to read and then I am fine with it.

Reply:
Yep fair 👍